src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,76 @@
+package com.google.step.snippet.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import java.io.IOException;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+@WebServlet("/vote")
+public class FeedbackServlet extends HttpServlet {
+
+  private static final String UP = "upvote";
+  private static final String DOWN = "downvote";
+  private static final String URL = "url";
+  private static final String FEEDBACK = "feedback";
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Entity feedbackEntity = null;
+    String url = request.getParameter(URL);
+
+    if (url == null) {
+      response.sendRedirect(request.getHeader("Referer"));
```

Comment:
It's bad practice to depend on user-supplied information, like request headers, without validating it. But I'm not sure what you're trying to do here - isn't this an async post request? I don't think you want to do any redirects.

Reply:
See #76 .

The request headers are being set by the POST request. I'm not sure what you mean by validating here beyond making null checks.